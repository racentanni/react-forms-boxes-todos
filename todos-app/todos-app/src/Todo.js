import React, { useState } from 'react';
import './Todo.scss';

function Todo({ id, task, removeTodo, updateTodo }) {
  const [isEditing, setIsEditing] = useState(false);
  const [newTask, setNewTask] = useState(task);
  const [isCompleted, setIsCompleted] = useState(false);

  const handleEdit = () => {
    setIsEditing(true);
  };

  const handleUpdate = (e) => {
    e.preventDefault();
    updateTodo(id, newTask);
    setIsEditing(false);
  };

  const handleToggleCompleted = () => {
    setIsCompleted(!isCompleted);
  };

  return (
    <div className={`Todo ${isCompleted ? 'completed' : ''}`}>
      {isEditing ? (
        <form onSubmit={handleUpdate}>
          <input
            type="text"
            value={newTask}
            onChange={(e) => setNewTask(e.target.value)}
          />
          <button type="submit">Save</button>
        </form>
      ) : (
        <>
          <span>{task}</span>
          <button onClick={handleEdit}>Edit</button>
          <button onClick={handleToggleCompleted}>
            {isCompleted ? 'Unmark' : 'Mark as completed'}
          </button>
          <button onClick={() => removeTodo(id)}>X</button>
        </>
      )}
    </div>
  );
}

export default Todo;