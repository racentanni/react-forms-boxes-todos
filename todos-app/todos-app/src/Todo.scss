.Todo {
    display: flex;
    align-items: center;
    justify-content: space-between;
    margin: 10px 0;
  
    &.completed span {
      text-decoration: line-through;
    }
  
    button {
      margin-left: 10px;
    }
  
    &.fade-out {
      animation: fadeOut 0.5s forwards;
    }
  }
  
  @keyframes fadeOut {
    from {
      opacity: 1;
    }
    to {
      opacity: 0;
    }
  }