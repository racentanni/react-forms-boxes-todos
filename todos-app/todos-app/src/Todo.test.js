import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Todo from './Todo';

test('renders without crashing', () => {
  render(<Todo id={1} task="Test Todo" removeTodo={() => {}} updateTodo={() => {}} />);
});

test('matches snapshot', () => {
  const { asFragment } = render(<Todo id={1} task="Test Todo" removeTodo={() => {}} updateTodo={() => {}} />);
  expect(asFragment()).toMatchSnapshot();
});

test('calls removeTodo on button click', () => {
  const removeTodo = jest.fn();
  const { getByText } = render(<Todo id={1} task="Test Todo" removeTodo={removeTodo} updateTodo={() => {}} />);
  const button = getByText('X');

  fireEvent.click(button);

  expect(removeTodo).toHaveBeenCalledWith(1);
});

test('can edit a todo', () => {
  const updateTodo = jest.fn();
  const { getByText, getByDisplayValue, getByRole } = render(<Todo id={1} task="Test Todo" removeTodo={() => {}} updateTodo={updateTodo} />);
  const editButton = getByText('Edit');

  fireEvent.click(editButton);

  const input = getByDisplayValue('Test Todo');
  fireEvent.change(input, { target: { value: 'Updated Todo' } });

  const saveButton = getByRole('button', { name: /save/i });
  fireEvent.click(saveButton);

  expect(updateTodo).toHaveBeenCalledWith(1, 'Updated Todo');
});

test('can mark a todo as completed', () => {
  const { getByText, rerender } = render(<Todo id={1} task="Test Todo" removeTodo={() => {}} updateTodo={() => {}} />);
  const markButton = getByText('Mark as completed');

  fireEvent.click(markButton);

  // Re-render the component to reflect the state change
  rerender(<Todo id={1} task="Test Todo" removeTodo={() => {}} updateTodo={() => {}} />);

  expect(getByText('Unmark')).toBeInTheDocument();
  expect(getByText('Test Todo')).toHaveStyle('text-decoration: line-through');
});