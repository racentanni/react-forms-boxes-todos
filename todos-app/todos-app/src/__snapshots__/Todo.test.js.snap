// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Todo "
  >
    <span>
      Test Todo
    </span>
    <button>
      Edit
    </button>
    <button>
      Mark as completed
    </button>
    <button>
      X
    </button>
  </div>
</DocumentFragment>
`;
